src/basic/virt.c

```diff
@@ -460,11 +460,30 @@ int detect_container(void) {
         _cleanup_free_ char *o = NULL;
         _cleanup_free_ char *p = NULL;
         const char *e = NULL;
+        unsigned i;
         int r;
 
+        static const struct {
+                const char *file_path;
+                int id;
+        } container_file_table[] = {
+                /* https://github.com/moby/moby/issues/18355 */
+                { "/.dockerenv",        VIRTUALIZATION_DOCKER },
```

Comment:
This is a bit tricky to do, because you still want Docker and (especially) Podman to override `container=oci` and `CONTAINER_OTHER`.

I think I got it working right in all the cases, but control flow got a little bit unwieldy, and I had to add an explicit check for `container=oci` which I previously avoided.